clang/lib/CodeGen/CodeGenFunction.cpp

```diff
@@ -1652,13 +1652,18 @@ void CodeGenFunction::GenerateCode(GlobalDecl GD, llvm::Function *Fn,
   if (getLangOpts().CUDA && !getLangOpts().CUDAIsDevice &&
       getLangOpts().SYCLIsHost && !FD->hasAttr<CUDAHostAttr>() &&
       FD->hasAttr<CUDADeviceAttr>()) {
-    Fn->setLinkage(llvm::Function::WeakODRLinkage);
     if (FD->getReturnType()->isVoidType())
       Builder.CreateRetVoid();
     else
       Builder.CreateRet(llvm::UndefValue::get(Fn->getReturnType()));
     return;
   }
+  // When compiling a CUDA file in SYCL device mode,
+  // set weak ODR linkage for possibly duplicated __device__ functions.
+  if (getLangOpts().CUDA && !getLangOpts().CUDAIsDevice &&
```

Comment:
Can you please explain why duplicate functions exist?

Reply:
Hi @elizabethandrews, thanks for your review.

Basically, there are two device compilations, one for **device-sycl** (0-3) and one for **device-cuda** (4-7),
https://github.com/intel/llvm/blob/cd67cf8566b928cba4fd1456b594db3a5b656180/clang/test/Driver/sycl-cuda-tu-offload.cu#L5-L13
 
If both the compilations generate the same function, the one from the **device-sycl** is marked as `weak_odr` for avoiding conflicts.

Reply:
Thanks for noticing it, @elizabethandrews!

I removed the check on CUDA attributes by mistake. The last commit re-introduces them.

Reply:
Is there a test we can add which could have caught this?

Reply:
Sure, the last commit updates test for checking it.